features/nvstore/source/nvstore.cpp

```diff
@@ -0,0 +1,926 @@
+
+static const uint32_t initial_crc = 0xFFFFFFFF;
+
+
+// -------------------------------------------------- Local Functions Declaration ----------------------------------------------------
+
+// -------------------------------------------------- Functions Implementation ----------------------------------------------------
+
+// Align a value to a specified size.
+// Parameters :
+// val           - [IN]   Value.
+// size          - [IN]   Size.
+// Return        : Aligned value.
+static inline uint32_t align_up(uint32_t val, uint32_t size)
+{
+    return (((val - 1) / size) + 1) * size;
+}
+
+// CRC32 calculation. Supports "rolling" calculation (using the initial value).
+// Parameters :
+// init_crc      - [IN]   Initial CRC.
+// data_size      - [IN]   Buffer's data size.
+// data_buf      - [IN]   Data buffer.
+// Return        : CRC.
+static uint32_t crc32(uint32_t init_crc, uint32_t data_size, uint8_t *data_buf)
+{
+    uint32_t i, j;
+    uint32_t crc, mask;
+
+    crc = init_crc;
+    for (i = 0; i < data_size; i++) {
+        crc = crc ^ (uint32_t) (data_buf[i]);
+        for (j = 0; j < 8; j++) {
+          mask = -(crc & 1);
+          crc = (crc >> 1) ^ (0xEDB88320 & mask);
+        }
+    }
+    return crc;
+}
+
+NVStore::NVStore() : _init_done(0), _init_attempts(0), _active_area(0), _max_keys(NVSTORE_MAX_KEYS),
+      _active_area_version(0), _free_space_offset(0), _size(0), _mutex(0), _offset_by_key(0), _flash(0),
+      _min_prog_size(0), _page_buf(0)
+{
+}
+
+NVStore::~NVStore()
+{
+    if (_init_done) {
+        deinit();
+    }
+}
+
+uint16_t NVStore::get_max_keys() const
+{
+    return _max_keys;
+}
+
+uint16_t NVStore::get_max_possible_keys()
+{
+    if (!_init_done) {
+        init();
+    }
+
+    size_t max_possible_keys = _size / align_up(sizeof(nvstore_record_header_t) * 2, _min_prog_size) - 1;
+
+    return (uint16_t)std::min(max_possible_keys, (size_t) last_reserved_key);
+}
+
+void NVStore::set_max_keys(uint16_t num_keys)
+{
+    MBED_ASSERT(num_keys < get_max_possible_keys());
+    _max_keys = num_keys;
+    // User is allowed to change number of keys. As this affects init, need to deinitialize now.
+    // Don't call init right away - it is lazily called by get/set functions if needed.
+    deinit();
+}
+
+int NVStore::flash_read_area(uint8_t area, uint32_t offset, uint32_t size, void *buf)
+{
+    return _flash->read(buf, _flash_area_params[area].address + offset, size);
+}
+
+int NVStore::flash_write_area(uint8_t area, uint32_t offset, uint32_t size, const void *buf)
+{
+    int ret;
+    // On some boards, write action can fail due to HW limitations (like critical drivers
+    // that disable all other actions). Just retry a few times until success.
+    for (int i = 0; i < num_write_retries; i++) {
+        ret = _flash->program(buf, _flash_area_params[area].address + offset, size);
+        if (!ret) {
+            return ret;
+        }
+#ifdef MBED_CONF_RTOS_PRESENT
+        rtos::Thread::wait(1);
```

Comment:
just calling wait for 1ms or greater will handle the rtos/non-rtos logic, it shouldn't be needed here.